tensorboard/webapp/metrics/views/card_renderer/card_lazy_loader.ts

```diff
@@ -160,20 +154,45 @@ export class CardLazyLoader implements OnInit, OnDestroy {
     private readonly store: Store<State>
   ) {}
 
-  onCardIntersection(enteredCards: Set<CardId>, exitedCards: Set<CardId>) {
+  onCardIntersection(
+    enteredElements: Set<Element>,
+    exitedElements: Set<Element>
+  ) {
+    const enteredCards = [...enteredElements].map((element) => {
+      const elementId = elementIdMap.get(element) ?? null;
+      const cardId = elementToCardIdMap.get(element) ?? null;
+      if (elementId === null || cardId === null) {
+        throw new Error(
+          'A CardObserver element must have an associated element id and card id.'
+        );
+      }
+      return {elementId, cardId};
+    });
+    const exitedCards = [...exitedElements].map((element) => {
+      const elementId = elementIdMap.get(element) ?? null;
+      const cardId = elementToCardIdMap.get(element) ?? null;
+      if (elementId === null || cardId === null) {
+        throw new Error(
+          'A CardObserver element must have an associated element id and card id.'
+        );
+      }
+      return {elementId, cardId};
+    });
     this.store.dispatch(
       actions.cardVisibilityChanged({enteredCards, exitedCards})
     );
   }
 
   ngOnInit() {
-    elementToCardIdMap.set(this.host.nativeElement, this.cardId);
+    const element = this.host.nativeElement;
+    elementToCardIdMap.set(element, this.cardId);
+    elementIdMap.set(element, nextElementId());
```

Comment:
Instead of two different Map, consider using a data structure like `{elementId: string; cardId: string}` then reading it would be much more easier. Logically, when cardId is null, elementId should never be null and as currently typed, this is hard to guarantee.

Reply:
Good idea, done.